Added a scrubbing step to the existing 3DRTMA workflow. This step is invoked at the start and clears COM and DATA directories that exceed their respective retention periods.

EdwardColon-NOAA opened the pull request:
The previous PR was accidentally intermingled with an older PR in my fork by repeating the branch name.

Commit:
Added a scrubbing step to the existing 3DRTMA workflow. This step is invoked at the start and clears COM and DATA directories that exceed their respective retention periods.

GangZhao-NOAA commented:
@EdwardColon-NOAA 
Hi Ed,
I didn't see Matthew's name in the reviewers list. You may want to add him.
Thanks,
-Gang

GangZhao-NOAA commented on workflow/rtma3d_pbspro_alaska.xml:
```diff
@@ -276,12 +278,22 @@
 
+<!ENTITY CLEAN_PROC "1">
+<!ENTITY CLEAN_RESOURCES '<walltime>00:30:00</walltime>'> 
+<!ENTITY CLEAN_RESERVATION '<nodes>1:ppn=128:mem=48GB</nodes><native>-l place=scatter</native><queue>&QUEUE;</queue><account>&ACCOUNT;</account>'>
```
Hi @EdwardColon-NOAA ,
Is this clean job a serial job or parallel job? does it need ppn=128 (128 PEs) to run? 
Thanks,
-Gang

GangZhao-NOAA commented on workflow/rtma3d_pbspro_conus.xml:
```diff
@@ -285,12 +288,23 @@
 
+<!ENTITY CLEAN_PROC "1">
+<!ENTITY CLEAN_RESOURCES '<walltime>00:30:00</walltime>'> 
+<!ENTITY CLEAN_RESERVATION '<nodes>1:ppn=128:mem=48GB</nodes><native>-l place=scatter</native><queue>&QUEUE;</queue><account>&ACCOUNT;</account>'>
```
Hi @EdwardColon-NOAA 
Same question for this ppn=128.
Thanks,
-Gang

GangZhao-NOAA commented on scripts/exrtma3d_clean.ksh:
```diff
@@ -0,0 +1,51 @@
+#!/bin/ksh --login
```
Hi @EdwardColon-NOAA (and @AnnetteGibbs-NOAA , @MatthewMorris-NOAA ),

"#!/bin/ksh --login", I remember that this "--login" is not encouraged by NCO. Not sure if what I remembered is correct.

Annette and Matthew, please correct me.

Thanks,
-Gang

EdwardColon-NOAA committed:
Update rtma3d_pbspro_conus.xml

Resolved resource allocation for clean step.

EdwardColon-NOAA committed:
Update rtma3d_pbspro_alaska.xml

Resolved resource allocation for clean step.

EdwardColon-NOAA committed:
Update exrtma3d_clean.ksh

Removed --login argument in environment call as its use is discouraged by NCO.

GangZhao-NOAA commented on workflow/rtma3d_pbspro_conus.xml:
```diff
@@ -285,12 +288,23 @@
 
+<!-- COM and DATA retention time (hours) -->
+
+<!ENTITY CLEAN_OLDCOM_HRS            "72">
```
@MatthewMorris-NOAA 
Hi Ed, this is not a question. Just a comment and actually for Matthew. 
Here, the CLEAN_OLDCOM_HRS=72, so 3 days old data under COM2 would be removed., including the autoqc data. Is it good for your autoqc step? I do not know how many days of history data your auqoqc requires? But if they are all saved in the database (*.db) file in the very previous cycle, then it should be OK to clean the data older than 3 days.

Thanks,

-Gang

MatthewMorris-NOAA replied:
Hi @GangZhao-NOAA,
As you've noted, the pertinent data is stored in the database (*db) files from the most recent cycle. However, the GSI ex-scripts are both set up to search back up to 7 days to find the most recent accept list. This was set up to account for machine outages, so I don't expect it to need to look back that far. I'd recommend increasing the COM retention somewhat (e.g., up to 5 days). DATA can be scrubbed earlier, but we want to give enough time to capture the data needed for any potential reruns.

EdwardColon-NOAA commented:
That is a good question, Gang. I will defer to Matt regarding the autoqc
retention period that might be needed.

-Edward

On Tue, Jun 3, 2025 at 2:02 PM Gang Zhao ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In workflow/rtma3d_pbspro_conus.xml
> <https://github.com/NOAA-EMC/NOAA_3drtma/pull/69#discussion_r2124549786>:
>
> >  <!-- Variables in Namelist -->
>  <!ENTITY GSI_grid_ratio_in_var       "1">
>  <!ENTITY GSI_grid_ratio_in_cldanl    "1">
>  <!ENTITY i_gsdsfc_uselist            "2">
>  <!ENTITY VERBOSE_GSI                 ".false.">
>
> +<!-- COM and DATA retention time (hours) -->
> +
> +<!ENTITY CLEAN_OLDCOM_HRS            "72">
>
> @MatthewMorris-NOAA <https://github.com/MatthewMorris-NOAA>
> Hi Ed, this is not a question. Just a comment and actually for Matthew.
> Here, the CLEAN_OLDCOM_HRS=72, so 3 days old data under COM2 would be
> removed., including the autoqc data. Is it good for your autoqc step? I do
> not know how many days of history data your auqoqc requires? But if they
> are all saved in the database (*.db) file in the very previous cycle, then
> it should be OK to clean the data older than 3 days.
>
> Thanks,
>
> -Gang
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/NOAA-EMC/NOAA_3drtma/pull/69#pullrequestreview-2893538414>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACFXXQHOS237UWCHQ6SMQPT3BXPHTAVCNFSM6AAAAAB6OVB4ISVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDQOJTGUZTQNBRGQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
-------------------------------------------------
Edward Colón
*Lynker* at NOAA/NWS/NCEP/EMC
5830 University Research Ct.
Rm #2025
Riverdale, MD 20737
Office: (301) 683-3815
Cell: (301) 213-3566

EdwardColon-NOAA commented:
Thanks for the input, Matt. I'll make that update.

-Edward

On Tue, Jun 3, 2025 at 2:13 PM MatthewMorris-NOAA ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In workflow/rtma3d_pbspro_conus.xml
> <https://github.com/NOAA-EMC/NOAA_3drtma/pull/69#discussion_r2124577992>:
>
> >  <!-- Variables in Namelist -->
>  <!ENTITY GSI_grid_ratio_in_var       "1">
>  <!ENTITY GSI_grid_ratio_in_cldanl    "1">
>  <!ENTITY i_gsdsfc_uselist            "2">
>  <!ENTITY VERBOSE_GSI                 ".false.">
>
> +<!-- COM and DATA retention time (hours) -->
> +
> +<!ENTITY CLEAN_OLDCOM_HRS            "72">
>
> Hi @GangZhao-NOAA <https://github.com/GangZhao-NOAA>,
> As you've noted, the pertinent data is stored in the database (*db) files
> from the most recent cycle. However, the GSI ex-scripts are both set up to
> search back up to 7 days to find the most recent accept list. This was set
> up to account for machine outages, so I don't expect it to need to look
> back that far. I'd recommend increasing the COM retention somewhat (e.g.,
> up to 5 days). DATA can be scrubbed earlier, but we want to give enough
> time to capture the data needed for any potential reruns.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/NOAA-EMC/NOAA_3drtma/pull/69#discussion_r2124577992>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACFXXQHDA7Y57UFN5ERJOZL3BXQV3AVCNFSM6AAAAAB6OVB4ISVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDQOJTGU4DAMRZHA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
-------------------------------------------------
Edward Colón
*Lynker* at NOAA/NWS/NCEP/EMC
5830 University Research Ct.
Rm #2025
Riverdale, MD 20737
Office: (301) 683-3815
Cell: (301) 213-3566

EdwardColon-NOAA committed:
Update rtma3d_pbspro_conus.xml

Increased the COM retention period to 5 days for autoQC requirements.

EdwardColon-NOAA committed:
Update rtma3d_pbspro_alaska.xml

Increased the COM retention period to 5 days for autoQC requirements.

AnnetteGibbs-NOAA commented on scripts/exrtma3d_clean.ksh:
```diff
@@ -0,0 +1,51 @@
+for onetime in ${XX[*]};do
+  if [[ ${onetime} =~ ^[0-9]+$ ]] && [[ ${onetime} -le ${deletetime} ]]; then
+    rm -rf ${DATAROOT}/pbspro/${NET}.${onetime}
+    echo "Deleted ${DATAROOT}/pbspro/${NET}.${onetime}"
```
It does not look like this is actually working.  Have you checked your logfiles?  Have you checked to see if it actually deleted the directories?  For instance, when I check your clean logfile for 2025060412, it says the following:

 0.006 + /lfs/h2/emc/da/noscrub/edward.colon/fork3/NOAA_3drtma/scripts/exrtma3d_clean.ksh
Deleting ptmp directories before 20250601...
Deleting stmp directories before 2025060112...
 0.527 + err=0
 0.527 + export err
 0.527 + err_chk
 completed cleanly

It does not say "Deleted.." and when I go to look at the DATA directory, this and preceding directories still exist:

drwxr-sr-x 15 edward.colon emc 4096 Jun  1 11:01 rtma3d.202506011000
drwxr-sr-x 15 edward.colon emc 4096 Jun  1 12:01 rtma3d.202506011100
drwxr-sr-x 16 edward.colon emc 4096 Jun  1 13:05 **rtma3d.202506011200**

Can you please double check that all is working correctly for both CONUS and Alaska, before I further review the PR.  Even though it says it is deleting the directories, it does not actually do it.

EdwardColon-NOAA commented:
Thanks Annette,

I’ll take another look. Previously, the algorithm seemed to be working.

-------------------------------------------------
Edward Colón
*Lynker* at NOAA/NWS/NCEP/EMC
5830 University Research Ct.
Rm #2025
Riverdale, MD 20737
Office: (301) 683-3815
Cell: (301) 213-3566


On Wed, Jun 4, 2025 at 10:02 AM Annette Gibbs ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In scripts/exrtma3d_clean.ksh
> <https://github.com/NOAA-EMC/NOAA_3drtma/pull/69#discussion_r2126683788>:
>
> > +    rm -rf ${COMROOT}/${NET}/pbspro/${NET}.${onetime}
> +    echo "Deleted ${COMROOT}/${NET}/pbspro/${NET}.${onetime}"
> +  fi
> +done
> +
> +#-----------------------------------------------------------------------
> +# Delete stmp directories
> +#-----------------------------------------------------------------------
> +deletetime=$(date +%Y%m%d%H -d "${currentime} ${CLEAN_OLDRUN_HRS} hours ago")
> +echo "Deleting stmp directories before ${deletetime}..."
> +cd ${DATAROOT}/pbspro
> +set -A XX $(ls -d ${NET}.20* | sort -r)
> +for onetime in ${XX[*]};do
> +  if [[ ${onetime} =~ ^[0-9]+$ ]] && [[ ${onetime} -le ${deletetime} ]]; then
> +    rm -rf ${DATAROOT}/pbspro/${NET}.${onetime}
> +    echo "Deleted ${DATAROOT}/pbspro/${NET}.${onetime}"
>
> It does not look like this is actually working. Have you checked your
> logfiles? Have you checked to see if it actually deleted the directories?
> For instance, when I check your clean logfile for 2025060412, it says the
> following:
>
> 0.006 +
> /lfs/h2/emc/da/noscrub/edward.colon/fork3/NOAA_3drtma/scripts/exrtma3d_clean.ksh
> Deleting ptmp directories before 20250601...
> Deleting stmp directories before 2025060112...
> 0.527 + err=0
> 0.527 + export err
> 0.527 + err_chk
> completed cleanly
>
> It does not say "Deleted.." and when I go to look at the DATA directory,
> this and preceding directories still exist:
>
> drwxr-sr-x 15 edward.colon emc 4096 Jun 1 11:01 rtma3d.202506011000
> drwxr-sr-x 15 edward.colon emc 4096 Jun 1 12:01 rtma3d.202506011100
> drwxr-sr-x 16 edward.colon emc 4096 Jun 1 13:05 *rtma3d.202506011200*
>
> Can you please double check that all is working correctly for both CONUS
> and Alaska, before I further review the PR. Even though it says it is
> deleting the directories, it does not actually do it.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/NOAA-EMC/NOAA_3drtma/pull/69#pullrequestreview-2896866285>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACFXXQBUA56HUATKDALBWUD3B332NAVCNFSM6AAAAAB6OVB4ISVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDQOJWHA3DMMRYGU>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

EdwardColon-NOAA committed:
Update exrtma3d_clean.ksh

A couple of additional changes to the DATA directory processing block were needed. The ex-script is now functional in removing directories older than the retention period.

EdwardColon-NOAA committed:
Update exrtma3d_clean.ksh

Added additional conditional blocks that check for the existence of COM and DATA directories before removing individual files coinciding with the retention period.

EdwardColon-NOAA committed:
Update exrtma3d_clean.ksh

Added additional log file messages that highlight which cycle is being scrubbed or, if a cycle in the COM or DATA directory does not exist because it was already scrubbed or the parallel has not run long enough, a log file message is also produced.

EdwardColon-NOAA committed:
Update exrtma3d_clean.ksh

Re-introduced conditional block that removes empty COM directories.
